### PR TITLE
Ported and adapted LedgerState from HORNET

### PR DESCRIPTION
**Pull request by alexsporn**
- Added simple import/export function
- Ported tests and added new ones for the new snapshot components

### CODE REVIEW

File: pkg/protocol/engine/ledgerstate/slot_diff.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+	return m.Bytes()
+}
+
+func (ms *SlotDiff) KVStorableKey() []byte {
+	return slotDiffKeyForIndex(ms.Index)
+}
+
+func (ms *SlotDiff) KVStorableValue() []byte {
+	m := marshalutil.New(9)
```
Comment (jonastheis) [inline pkg/protocol/engine/ledgerstate/slot_diff.go:37]: This serialization stuff we might be able to simplify a bit using serix and maybe a typed KV store. But no need to do it now imo.

**Review by jonastheis — APPROVED** [review]

File: pkg/protocol/engine/ledgerstate/manager.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	}
+
+	if total != tokenSupply {
+		return ErrOutputsSumNotEqualTotalSupply
+	}
+
+	return nil
+}
+
+func (m *Manager) AddUnspentOutputWithoutLocking(unspentOutput *Output) error {
```
[the remark this answers is not in the document]
Comment (alexsporn) [reply pkg/protocol/engine/ledgerstate/manager.go:264]: This is used to add outputs for creating a snapshot. Or when importing the ledger from a snapshot, you add all the unspent outputs, set the ledger index and then can appy/rollback diffs

File: pkg/protocol/engine/ledgerstate/manager.go [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+}
+
+func (m *Manager) ReadLedgerIndex() (iotago.SlotIndex, error) {
+	m.ReadLockLedger()
+	defer m.ReadUnlockLedger()
+
+	return m.ReadLedgerIndexWithoutLocking()
+}
+
+func (m *Manager) ApplyConfirmationWithoutLocking(index iotago.SlotIndex, newOutputs Outputs, newSpents Spents) error {
```
[the remark this answers is not in the document]
Comment (alexsporn) [reply pkg/protocol/engine/ledgerstate/manager.go:139]: Sure, I can rename it. The names come from HORNET

File: pkg/protocol/engine/ledgerstate/database_prefixes.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+       40 bytes  +    8 bytes   +             8 bytes           +   1 byte type + X bytes
+
+   Spent Output:
+   ================
+   Key:
+       StoreKeyPrefixSpent + iotago.OutputID
+            1 byte         +     34 bytes
+
+   Value:
+       TargetTransactionID (iotago.TransactionID) + ConfirmationIndex (iotago.SlotIndex) + ConfirmationTimestamp (time.Time)
```
[the remark this answers is not in the document]
Comment (alexsporn) [reply pkg/protocol/engine/ledgerstate/database_prefixes.go:47]: This is actually the spent transaction acceptance index. Let me check how to properly rephrase it

**Pull request merged**